### ISSUE TITLE
Performance of SurfacePlotMesh for complex function

### ISSUE THREAD

**Issue by ThomasDaheim** [FXyz/FXyz#150]
After some playing around I managed to get SurfacePlotMesh to work and to show a landscape based on SRTM height data:

![image](https://user-images.githubusercontent.com/14020617/149838389-4a1261fe-10a7-4353-aae6-78336a1bae1e.png)

Unfortunately, the function to determine the height per location is costly. So when trying to find ways to speed up things I found the following code in SurfacePlotMesh  the seems to indicate that the function is called twice for every point:

createPlotMesh:
```
        // Create points
        for (int y = 0; y <= divisionsY; y++) {
            float dy = (float)(-rangeY/2d + ((float)y /(float)divisionsY)*rangeY);
            for (int x = 0; x <= divisionsX; x++) {
                float dx = (float)(-rangeX/2d + ((float)x /(float)divisionsX)*rangeX);
                pointY = (float)scale*function2D.apply(new Point2D(dx,dy)).floatValue();
                listVertices.add(new Point3D(dx, pointY, dy));
            }
        }
```

createPolygonMesh:
```
        // Create points
        float[] points = new float[(divisionsY + 1) * numDivX * 3];
        int counter = 0;
        for (int y = 0; y <= divisionsY; y++) {
            float dy = (float)(-rangeY/2d + ((float)y /(float)divisionsY)*rangeY);
            for (int x = 0; x <= divisionsX; x++) {
                float dx = (float)(-rangeX/2d + ((float)x /(float)divisionsX)*rangeX);
                pointY = 1.0f*(float)scale*function2D.apply(new Point2D(dx,dy)).floatValue();
                points[counter++] = dx;
                points[counter++] = pointY;
                points[counter++] = dy;
            }
        }
```

Seems like those functions are never called separately, only together in `updateMesh()`.

Is that correct? If yes, it should be possible to eliminate one of the calls?

**Comment by jperedadnr** [comment]
That pic looks awesome!

You are right, we are creating two different meshes based on the same 2D grid, and the first call creates `listVertices`, that could be used for the second. 

However, the need for polygonMesh is also in question here, as it is only required if isWireframe is activated.

Definitely, there is room for improvement. Do you want to give it a try and contribute a PR?

**Comment by ThomasDaheim** [comment]
Thanks for confirming my understanding :-)

As for a PR: A have created a local fast version of SurfacePlotMesh that removes this (and other) duplications. But let my try and work more with the library before coming up with a PR. There are also a view things I would like to understand first:

- add add. lines to the plot (tracks I took with a GPS)
- add axis description and legends

After that I can do with FXyz all I previously did with jzy3d and hopefully understand the code a lot better...

**Comment by Birdasaur** [comment]
Agreed with Jose, very cool. JavaFX really shines when you have enough data like the STRM sets you have.

**Comment by ThomasDaheim** [comment]
By now I have all functionality from jzy3d also with FXyz :-) Before thinking about a pull request I was wandering what kind of test cases I should check my modifications against? I can't seem to find any test cases using SurfacePlotMesh?

![Clipboard02](https://user-images.githubusercontent.com/14020617/180074409-206f1f9f-2f09-468f-8a8b-6fec1377731e.jpg)

**Comment by Birdasaur** [comment]
Hi Thomas. Unfortunately we don't really have any performance or functionality tests (that I can think of) for the SurfacePlotMesh.
The best I can come up with is ensuring that it renders as expected in the Sampler App (sampler module).
That said between Jose and I we'd be happy to review your proposed changes.  I'm currently using a modified version of the SurfacePlotMesh myself in an application and could test it against some of the stuff I'm doing. (unfortunately I can't share that stuff... yet.)

**Comment by ThomasDaheim** [comment]
Thanks fpr the info. Will try to make my changes run with the Sampler App.

Unfortunately, I run into `org.joor.ReflectException: java.lang.NoSuchFieldException: javaExecHandleBuilder` exception when trying to run it. I'm using:

- Win10
- Netbeans 14
- AdoptOpenJDK\jdk-15.0.2.7-hotspot"
- gradle-6.9.1

Google found a few hints for this exception but nothing particularly useful. Have you maybe encountered this before?

**Comment by Birdasaur** [comment]
We just pushed into Master a bunch of upgrades from @jperedadnr that bump up both Gradle and the JDK to 7.6 and 17 respectively, along upgraded versions of various dependencies and plugins in the build script. 
Feel free to pull and see if you still get these issues. I'm using Win10, Netbeans 15 running on OpenJDK 18
